helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroRecordUtil.java

```diff
@@ -811,4 +820,147 @@ private static Enum<?> getSpecificEnumSymbol(Class<? extends Enum<?>> enumClass,
       throw new IllegalStateException("while trying to resolve " + enumClass.getName() + ".valueOf(" + symbolStr + ")", e);
     }
   }
+
+  /**
+   * verifies that the schemas on any nested named types "contained" by a given IndexedRecord
+   * match the schemas declared by the records schema (or SCHEMA$ field for specific record classes).
+   * this is useful to validate in situations where various generated record classes can come from
+   * different jars (and so could have a schema different than the one contained in the outermost
+   * record's schema)
+   * @param record a record, which could contain other records or named types
+   * @throws com.linkedin.avroutil1.compatibility.exception.InconsistentSchemaException if the schema for the record does not match the schema declared by any nested object
+   */
+  public static void validateNestedSchemasConsistent(IndexedRecord record) throws Exception {
+    if (record == null) {
+      return;
+    }
+    Schema outermost = record.getSchema();
+    Map<String, Schema> referenceDefinitions = AvroSchemaUtil.getAllDefinedSchemas(outermost);
+    AvroPath path = new AvroPath();
+    IdentityHashMap<Schema, Boolean> visited = new IdentityHashMap<>();
+    SchemaComparisonConfiguration comparisonConfiguration = SchemaComparisonConfiguration.STRICT;
+    //noinspection deprecation
+    if (AvroCompatibilityHelper.getRuntimeAvroVersion().earlierThan(AvroVersion.AVRO_1_8)) {
+      //avro < 1.7.3 cant handle non-string properties, so we use a less strict comparison
+      comparisonConfiguration = SchemaComparisonConfiguration.PRE_1_7_3;
+    }
+    validateNestedRecordSchema(record, path, visited, referenceDefinitions, comparisonConfiguration);
+  }
+
+  private static void validateNestedRecordSchema(
+          IndexedRecord record,
+          AvroPath path,
+          IdentityHashMap<Schema, Boolean> visited,
+          Map<String, Schema> referenceDefinitions,
+          SchemaComparisonConfiguration comparisonConfiguration
+  ) throws InconsistentSchemaException {
+    Schema recordSchema = record.getSchema();
+
+    //validate current record's schema if we havent before
```

Comment:
`if we haven't before`...where's the logic to only validate if we haven't before happening?

Reply:
Oh it's in the `validateNestedSchema()` call. So we recurse through the whole record but only check the schema if we haven't checked it before

Reply:
added doc to the declaration ofthe "visited" map to explain this